### PR TITLE
Feature CMR-7760 - Adding Support for Metadata Specification and Related URL to UMM-Var

### PR DESCRIPTION
**Pull request by jceaser**
* Adding UMM-Var 1.8 support
* Adding Related URL field
* Adding Metadata Specification field
* Related URL fields Mime Type, Format, and Content Type/Type/Subtype are validated against KMS
* UMM-Var will always return errors for Content-Type/Type/Subtype

### CODE REVIEW

File: umm-spec-lib/resources/json-schemas/variable/umm/v1.8/umm-var-1.8-sample.json [hunk 1]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (eereiter) [inline umm-spec-lib/resources/json-schemas/variable/umm/v1.8/umm-var-1.8-sample.json:1]: This file should go under the example_data directory and not with the schema.

**Review by eereiter — CHANGES_REQUESTED** [review]
Looks like 1 spacing issue.  Siwei corrected me on the others. 
Is the sample file in the schema being used as a test?  I would deleted if not, otherwise move it to the example_data/json_files/version directory.

File: umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj [hunk 2]
```diff
@@ -1,17 +1,24 @@
...
+  {:collection "UMM-C"  ; Currently does not have this node
    :granule "UMM-G"
    :service "UMM-S"
    :tool "UMM-T"
-   :variable "UMM-V"})  ; currently does not have this node
+
+   ;; UMM-V can not be used due to visualizations. Concept IDs start with V.
+   :variable "UMM-Var"
+
+   ;; Currently not implimented but this name is reserved for future use.
```
Comment (jaybarra) [inline umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:19]: typo: implemented

**Review by eereiter — APPROVED** [review]

File: umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj [hunk 3]
```diff
@@ -1,17 +1,24 @@
...
-   functions is needed."
+   functions is needed.
+   "
   (:require
    [cmr.umm-spec.versioning :as ver]))
 
 (def types
   "This is a mapping between collection symbol types to names found in metadata"
-  {:collection "UMM-C"  ; currently does not have this node
+  {:collection "UMM-C"  ; Currently does not have this node
```
Comment (jaybarra) [inline umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:11]: Is that comment still applicable?

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 4]
```diff
@@ -183,36 +180,42 @@
...
-                             msg/getdata-format-not-matches-kms-keywords)
-                  :files (v/every {:format (match-getdata-format-kms-keywords-validation
-                                             kms-index
-                                             msg/getdata-format-not-matches-kms-keywords)})}}))
+      :data-granule {:format (match-kms-keywords-validation-single
+                              kms-index
+                              :granule-data-format
+                              msg/getdata-format-not-matches-kms-keywords)
+                     :files (v/every {:format (match-kms-keywords-validation-single
+                                               kms-index
```
[the remark this answers is not in the document]
Comment (jceaser) [reply ingest-app/src/cmr/ingest/validation/validation.clj:216]: I would like to go with what ever the editor does for me when I hit enter

File: umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj (hunk 3 above)
Comment (jceaser) [reply umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:11]: it is still correct, I don't know where it is on the schedule, but we are running out of small schemas to change, so it's day is coming

Comment (jceaser) [reply umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:11]: I'm adding the word "Reserved" to the comment.

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 5]
```diff
@@ -383,12 +386,32 @@
...
     {:MeasurementIdentifiers (measurement-validation
-                              kms-index msg/measurements-not-matches-kms-keywords)}))
+                              kms-index msg/measurements-not-matches-kms-keywords)
+
+     :RelatedURLs (v/every [{:Format (match-kms-keywords-validation-single
+                                      kms-index
+                                      :granule-data-format
+                                      msg/getdata-format-not-matches-kms-keywords)}
+                            {:MimeType (match-kms-keywords-validation-single
+                                        kms-index
```
[the remark this answers is not in the document]
Comment (jceaser) [reply ingest-app/src/cmr/ingest/validation/validation.clj:396]: No change, all lines where Atom wants them

File: ingest-app/src/cmr/ingest/validation/validation.clj (hunk 4 above)
[the remark this answers is not in the document]
Comment (jceaser) [reply ingest-app/src/cmr/ingest/validation/validation.clj:216]: I checked all of @eereiter suggestions against where Atom wants things. Some lines changed, some did not. I would like to resolve this thread based on deferring to Atom and will do so latter today unless I hear a complaint.

File: umm-spec-lib/resources/json-schemas/variable/umm/v1.8/umm-var-1.8-sample.json (hunk 1 above)
Comment (jceaser) [reply umm-spec-lib/resources/json-schemas/variable/umm/v1.8/umm-var-1.8-sample.json]: file removed
TL;DR
File deleted after manual inspection of test code and I did a grep for the file name resulting in no usage.

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 6]
```diff
@@ -183,36 +180,42 @@
...
+                                                :granule-data-format
                                                 msg/getdata-format-not-matches-kms-keywords)})})}}))
 
 (defn granule-keyword-validations
   "These are the keyword validation rules needed for granule metadata. Remember
    granules are in the legacy format."
   [context]
   (let [kms-index (kms-fetcher/get-kms-index context)]
-    {:RelatedUrls {:related-urls (v/every {:format (match-getdata-format-kms-keywords-validation
+    {:RelatedUrls {:related-urls (v/every {:format (match-kms-keywords-validation-single
```
Comment (daniel-zamora) [inline ingest-app/src/cmr/ingest/validation/validation.clj:197]: Indentation looks off.

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 7]
```diff
@@ -183,36 +180,42 @@
...
                                  kms-index
+                                 :granule-data-format
                                  msg/getdata-format-not-matches-kms-keywords)})}
      :DataGranule {:ArchiveAndDistributionInformation {
-                   :Format (match-getdata-format-kms-keywords-validation
+                   :Format (match-kms-keywords-validation-single
                              kms-index
```
Comment (daniel-zamora) [inline ingest-app/src/cmr/ingest/validation/validation.clj:203]: Indentation looks off.

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 8]
```diff
@@ -383,12 +386,32 @@
...
+     :RelatedURLs (v/every [{:Format (match-kms-keywords-validation-single
+                                      kms-index
+                                      :granule-data-format
+                                      msg/getdata-format-not-matches-kms-keywords)}
+                            {:MimeType (match-kms-keywords-validation-single
+                                        kms-index
+                                        :mime-type
+                                        msg/mime-type-not-matches-kms-keywords)}])}))
+
+(defn- variable-keyword-validations-non-ignorable
```
Comment (daniel-zamora) [inline ingest-app/src/cmr/ingest/validation/validation.clj:400]: unignorable?

File: ingest-app/src/cmr/ingest/validation/validation.clj [hunk 9]
```diff
@@ -383,12 +386,32 @@
...
+  [context]
+  (let [kms-index (kms-fetcher/get-kms-index context)]
+    {:RelatedURLs [(match-related-url-kms-keywords-validations kms-index)]}))
 
 (defn umm-spec-validate-variable
   "Validate variable through umm-spec validation functions. If warn? flag is
   true and umm-spec-validation is off, log warnings and return messages, otherwise throw errors."
   [variable context warn?]
+  (when-let [non-ignorable (seq (umm-spec-validation/validate-variable-with-no-defaults
+                                 variable [(variable-keyword-validations-non-ignorable context)]))]
```
Comment (daniel-zamora) [inline ingest-app/src/cmr/ingest/validation/validation.clj:412]: Better formatting would be:
```
 (when-let [non-ignorable (seq (umm-spec-validation/validate-variable-with-no-defaults
                                variable 
                                [(variable-keyword-validations-non-ignorable context)]))]
```

File: umm-spec-lib/src/cmr/umm_spec/migration/version/variable.clj [hunk 10]
```diff
@@ -210,3 +211,17 @@
       (assoc-in [:Characteristics :IndexRanges] (get v :IndexRanges))
       (assoc-in [:Characteristics :GroupPath] (get v :Name))
       (dissoc :IndexRanges :StandardName)))
+
+;; migrations for 1.8 **********************************************************
+
+(defmethod interface/migrate-umm-version [:variable "1.7" "1.8"]
+  [context umm-v & _]
+  ;insert a metadata specification
```
Comment (daniel-zamora) [inline umm-spec-lib/src/cmr/umm_spec/migration/version/variable.clj:219]:
``` 
; insert a metadata specification
```

File: umm-spec-lib/src/cmr/umm_spec/migration/version/variable.clj [hunk 11]
```diff
@@ -210,3 +211,17 @@
...
+  (-> umm-v
+      (m-spec/update-version :variable "1.8")))
+
+(defmethod interface/migrate-umm-version [:variable "1.8" "1.7"]
+  ;drop metadata specification and related urls
```
Comment (daniel-zamora) [inline umm-spec-lib/src/cmr/umm_spec/migration/version/variable.clj:224]:
```  
; drop metadata specification and related urls
```

File: umm-spec-lib/test/cmr/umm_spec/test/migration/version/variable.clj [hunk 12]
```diff
@@ -12,6 +12,9 @@
    [cmr.umm-spec.versioning :as v]
    [com.gfredericks.test.chuck.clojure-test :refer [for-all]]))
 
+;;All migrations in this file will migrate variables in this way
```
Comment (daniel-zamora) [inline umm-spec-lib/test/cmr/umm_spec/test/migration/version/variable.clj:15]:
```
;; All migrations in this file will migrate variables in this way
```

File: ingest-app/src/cmr/ingest/validation/validation.clj (hunk 8 above)
Comment (jceaser) [reply ingest-app/src/cmr/ingest/validation/validation.clj:400]: I can try that, I have not been happy with this name and welcome suggestions.

**Review by sxu123 — APPROVED** [review]

File: umm-spec-lib/resources/json-schemas/variable/umm/v1.8/business_rules_v1.8.md [hunk 13]
```diff
@@ -0,0 +1,14 @@
+## Changes for 1.8
+
+(This is accompanying ancillary documenation that is useful for CMR and MMT to handle the latest implementation of schema changes.  These include any noteworthy ingest/field migration rules, how the MMT forms should handle the new/additional fields, how the CMR preview should handle the new/additional fields, etc.).
+
+### CMR
+- Migrating UMM-Var records up to 1.8 add the MetadataSpecification for version 1.8
+  Migrate down to 1.7 from 1.8 drop MetadataSpecification drop RelatedURLs.
```
Comment (jaybarra) [inline umm-spec-lib/resources/json-schemas/variable/umm/v1.8/business_rules_v1.8.md:7]: wording for the end of the sentence is odd, does it drop both MetadataSpecificiation and RelatedURLs?

File: umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj [hunk 14]
```diff
@@ -1,17 +1,24 @@
 (ns cmr.umm-spec.metadata-specification
   "Functions related to the MetadataSpecification node common to many of the umm
    models. Eventually all models will include this node, so a common set of
-   functions is needed."
+   functions is needed.
+   "
```
Comment (jaybarra) [inline umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:5]: You can end the comment on the same line, newline isn't necessary here

**Review by jaybarra — APPROVED** [review]

File: umm-spec-lib/resources/json-schemas/variable/umm/v1.8/business_rules_v1.8.md (hunk 13 above)
Comment (jceaser) [reply umm-spec-lib/resources/json-schemas/variable/umm/v1.8/business_rules_v1.8.md:7]: yes, but I merged already. Will correct in my next change.

File: umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj (hunk 14 above)
Comment (jceaser) [reply umm-spec-lib/src/cmr/umm_spec/metadata_specification.clj:5]: Sorry, I merged already, will add to my next change.
